### PR TITLE
[grid] Improve SlotMatcher and SlotSelector on request browserVersion

### PR DESCRIPTION
**Pull request by VietND96**
### **User description**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description
This improvement helps autoscaling control strictly on a request without the capability to set `browserVersion` will be assigned to list Slots with top priority on stereotype without `browserVersion` or a latest `browserVersion` (based on SemVer comparison).

Improve DefaultSlotMatcher, where the request with cap `browserVersion=130` can match to slot with stereotype `browserVersion=130.0` (based on SemVer comparison return 0). Few corner case also covered e.g

```
    assertThat(NodeStatus.semVerComparator("131.0.6778.85", "131")).isEqualTo(0); -> match
    assertThat(NodeStatus.semVerComparator("131.0.6778.85", "131.0")).isEqualTo(0); -> match
    assertThat(NodeStatus.semVerComparator("131.0.6778.85", "131.0.6778")).isEqualTo(0); -> match
    assertThat(NodeStatus.semVerComparator("131.0.6778.85", "131.0.6778.95")).isEqualTo(1); -> not match
```

### Motivation and Context
This pull request introduces several important changes to the Selenium Grid project, focusing on improving the way browser versions are compared and sorted. The changes include implementing a custom semantic version comparator, updating the slot selection logic, and adding relevant tests.

### Improvements to browser version comparison:

* [`java/src/org/openqa/selenium/grid/data/NodeStatus.java`](diffhunk://#diff-04bf71dc8f94e5e1e3620bb9b03ea6162519e7adefa5dbec7a420a5c88dfc1d8R207-R265): Added a `getBrowserVersion` method and a custom `semVerComparator` method for comparing semantic versions with empty strings considered first.

### Updates to slot selection logic:

* [`java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java`](diffhunk://#diff-08612f5a21ea03efdd38ead8885dd4b72be486ea76f0de8f0a4dce609f34af0dL85-R87): Modified the `matches` method to use the new `semVerComparator` for browser version comparison.
* [`java/src/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelector.java`](diffhunk://#diff-07c6f048022f33fe17f8323da83d55e0a9e09ecd010a4d3fcafd25ad62f8dbe4R56-R60): Updated the slot selection logic to sort nodes by browser version using the new comparator.

### Addition of tests:

* [`java/test/org/openqa/selenium/grid/data/NodeStatusTest.java`](diffhunk://#diff-593943d5e5fd01f7d03d2e32f8aea53c9d25d46996296e231b2afb1eb19925edR37-R57): Added tests for the `semVerComparator` method to ensure correct browser version matching.
* [`java/test/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelectorTest.java`](diffhunk://#diff-c3acffd6266dc62129064038e49aea1f6dfaf453dbd0aac738d281697babc892R90-R126): Added tests to verify nodes are ordered correctly by browser version. [[1]](diffhunk://#diff-c3acffd6266dc62129064038e49aea1f6dfaf453dbd0aac738d281697babc892R90-R126) [[2]](diffhunk://#diff-c3acffd6266dc62129064038e49aea1f6dfaf453dbd0aac738d281697babc892R294-R307)

### Miscellaneous:

* [`rust/src/lock.rs`](diffhunk://#diff-f71dffcd0aff03c2e0e4adf54f9334ec319eb3051c7edd1f721480fe338e46afR1-R17): Added license information to the file.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

___

### **PR Type**
Enhancement

___

### **Description**
Enhanced browser version handling in Selenium Grid with the following improvements:

- Implemented semantic version comparison for better browser version matching
  * Supports partial version matching (e.g., "131" matches "131.0.6778.85")
  * Handles empty versions and special versions (e.g., "stable", "beta")
  * Provides descending order comparison for version sorting

- Improved slot selection and matching logic:
  * Added browser version as a sorting criterion in slot selection
  * Prioritizes slots with empty browser versions
  * Enhanced version comparison in DefaultSlotMatcher

- Added comprehensive test coverage:
  * Test cases for various version formats and edge cases
  * Tests for node ordering based on browser versions
  * Verification of slot selection behavior

- Code quality improvements:
  * Added missing license header to Rust file
  * Enhanced code documentation

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>DefaultSlotMatcher.java</strong><dd><code>Improved browser version matching logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java

<li>Enhanced browser version matching using semantic version comparison<br> <li> Replaced direct version equality check with <code>semVerComparator</code> for more <br>flexible matching<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14914/files#diff-08612f5a21ea03efdd38ead8885dd4b72be486ea76f0de8f0a4dce609f34af0d">+3/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>NodeStatus.java</strong><dd><code>Added semantic version comparison functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/src/org/openqa/selenium/grid/data/NodeStatus.java

<li>Added new <code>getBrowserVersion</code> method to retrieve minimum browser version<br> <li> Implemented custom <code>semVerComparator</code> for semantic version comparison<br> <li> Added helper method <code>isNumber</code> for version parsing<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14914/files#diff-04bf71dc8f94e5e1e3620bb9b03ea6162519e7adefa5dbec7a420a5c88dfc1d8">+59/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>DefaultSlotSelector.java</strong><dd><code>Enhanced slot selection with version-based sorting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/src/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelector.java

<li>Added browser version comparison to slot selection sorting criteria<br> <li> Enhanced slot selection to prioritize empty versions first<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14914/files#diff-07c6f048022f33fe17f8323da83d55e0a9e09ecd010a4d3fcafd25ad62f8dbe4">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>NodeStatusTest.java</strong><dd><code>Added tests for semantic version comparison</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/test/org/openqa/selenium/grid/data/NodeStatusTest.java

<li>Added comprehensive test cases for browser version comparison<br> <li> Covered various version formats and edge cases<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14914/files#diff-593943d5e5fd01f7d03d2e32f8aea53c9d25d46996296e231b2afb1eb19925ed">+21/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>DefaultSlotSelectorTest.java</strong><dd><code>Added tests for version-based node ordering</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/test/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelectorTest.java

<li>Added test cases for node ordering by browser version<br> <li> Added helper method for creating nodes with specific stereotypes<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14914/files#diff-c3acffd6266dc62129064038e49aea1f6dfaf453dbd0aac738d281697babc892">+51/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>lock.rs</strong><dd><code>Added missing license header</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rust/src/lock.rs

- Added license header to the file

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14914/files#diff-f71dffcd0aff03c2e0e4adf54f9334ec319eb3051c7edd1f721480fe338e46af">+17/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: java/src/org/openqa/selenium/grid/data/NodeStatus.java [hunk 1]
```diff
@@ -204,6 +204,65 @@ public long getLastSessionCreated() {
...
+    if (v1.isEmpty()) return -1; // Empty string comes first
+    if (v2.isEmpty()) return 1;
+
+    String[] parts1 = v1.split("\\.");
+    String[] parts2 = v2.split("\\.");
+
+    int maxLength = Math.max(parts1.length, parts2.length);
+    for (int i = 0; i < maxLength; i++) {
+      String part1 = i < parts1.length ? parts1[i] : "0";
+      String part2 = i < parts2.length ? parts2[i] : part1;
```
Comment (diemol) [inline java/src/org/openqa/selenium/grid/data/NodeStatus.java]: This code shouldn't be here. It probably belongs to the `DefaultSlotMatcher`.

File: java/src/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelector.java [hunk 2]
```diff
@@ -53,6 +53,11 @@ public Set<SlotId> selectSlot(
                 .thenComparingDouble(NodeStatus::getLoad)
                 // Then last session created (oldest first), so natural ordering again
                 .thenComparingLong(NodeStatus::getLastSessionCreated)
+                // Then sort by stereotype browserVersion (descending order). SemVer comparison with
+                // considering empty value at first.
+                .thenComparing(
+                    Comparator.comparing(
+                        NodeStatus::getBrowserVersion, NodeStatus::semVerComparator))
```
Comment (diemol) [inline java/src/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelector.java]: The `semVerComparator` returns `true` or `false`. This is more suitable for a filter. Which is done below, through the matcher.

**Review by diemol — CHANGES_REQUESTED** [review]

File: java/src/org/openqa/selenium/grid/data/NodeStatus.java (hunk 1 above)
Comment (VietND96) [reply java/src/org/openqa/selenium/grid/data/NodeStatus.java]: I moved it to `CapabilitiesComparator` for reusable. Since this is used for both sorting in selector and slot matcher

File: java/src/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelector.java (hunk 2 above)
Comment (VietND96) [reply java/src/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelector.java]: This is still needed since multiple matches returned from SlotMatcher should be sorted to move the latest version to top

File: java/src/org/openqa/selenium/grid/data/CapabilitiesComparator.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.grid.data;
+
+public class CapabilitiesComparator {
```
Comment (joerg1985) [inline java/src/org/openqa/selenium/grid/data/CapabilitiesComparator.java]: I think this could implemenet `java.util.Comparator` and the name could be something like `SemanticVersionComparator`

File: java/src/org/openqa/selenium/grid/data/CapabilitiesComparator.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+      boolean isPart1Numeric = isNumber(part1);
+      boolean isPart2Numeric = isNumber(part2);
+
+      if (isPart1Numeric && isPart2Numeric) {
+        // Compare numerically
+        int num1 = Integer.parseInt(part1);
+        int num2 = Integer.parseInt(part2);
+        if (num1 != num2) {
+          return Integer.compare(num2, num1); // Descending order
```
Comment (joerg1985) [inline java/src/org/openqa/selenium/grid/data/CapabilitiesComparator.java]: Ascending order is the common order for `Comparator`s in Java, if they need to be decending just call `Comparator.reversed()`.

File: java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java [hunk 5]
```diff
@@ -91,6 +92,10 @@ public boolean matches(Capabilities stereotype, Capabilities capabilities) {
     return browserNameMatch && browserVersionMatch && platformNameMatch;
   }
 
+  private Boolean browserVersionMatch(String stereotype, String capabilities) {
```
Comment (joerg1985) [inline java/src/org/openqa/selenium/grid/data/DefaultSlotMatcher.java]: not realy relevant, just as an info: this could be a primitive `boolean`, not an `Boolean`.

File: java/src/org/openqa/selenium/grid/data/NodeStatus.java [hunk 6]
```diff
@@ -204,6 +204,14 @@ public long getLastSessionCreated() {
         .orElse(0);
   }
 
+  public String getBrowserVersion() {
+    return slots.parallelStream()
```
Comment (joerg1985) [inline java/src/org/openqa/selenium/grid/data/NodeStatus.java]: No need for offloading to different threads here, this might add latency to spawn the threads.

File: java/test/org/openqa/selenium/grid/data/DefaultSlotMatcherTest.java [hunk 7]
```diff
@@ -29,6 +30,27 @@ class DefaultSlotMatcherTest {
...
+    assertThat(browserVersionComparator("", "130.0")).isEqualTo(-1);
+    assertThat(browserVersionComparator("131.0.6778.85", "131")).isEqualTo(0);
+    assertThat(browserVersionComparator("131.0.6778.85", "131.0")).isEqualTo(0);
+    assertThat(browserVersionComparator("131.0.6778.85", "131.0.6778")).isEqualTo(0);
+    assertThat(browserVersionComparator("131.0.6778.85", "131.0.6778.95")).isEqualTo(1);
+    assertThat(browserVersionComparator("130.0", "130.0")).isEqualTo(0);
+    assertThat(browserVersionComparator("130.0", "130")).isEqualTo(0);
+    assertThat(browserVersionComparator("130.0.1", "130")).isEqualTo(0);
+    assertThat(browserVersionComparator("130.0.1", "130.0.1")).isEqualTo(0);
+    assertThat(browserVersionComparator("133.0a1", "133")).isEqualTo(0);
```
Comment (joerg1985) [inline java/test/org/openqa/selenium/grid/data/DefaultSlotMatcherTest.java]: A common test for comperators is to flip the input value to see if the compare does return the reversed (or in this case same) value for both calls.
e.g. `assertThat(browserVersionComparator("133", "133.0a1")).isEqualTo(0);` does fail.
This might be important, the DefaultSlotSelector will try to sort the slots and the input order is unspecified.

It might be best to only compare the minimum length of the arrays resulting from splitting at '.'

**Review by joerg1985 — COMMENTED** [review]
I have left some ideas for improvments, i hope this helps.

**Comment by VietND96** [comment]
Thanks @joerg1985, can you check is it better now?

**Comment by joerg1985** [comment]
@VietND96 i still think that `assertThat(comparator.compare("133", "133.0a1"))` should return 0.
Otherwise you will have effects like this one: 
```
    List<String> a = new java.util.ArrayList<>(List.of("133", "133.0a1"));
    List<String> b = new java.util.ArrayList<>(List.of("133.0a1", "133"));

    a.sort(new org.openqa.selenium.grid.data.SemanticVersionComparator());
    b.sort(new org.openqa.selenium.grid.data.SemanticVersionComparator());

    assertThat(a).isEqualTo(b);
```

**Comment by joerg1985** [comment]
@VietND96 i think this comparator is not suitable for sorting the node slots.
`133.0a1` is dummy data so we might just ignore it...

 e.g.:
```
    List<String> a = new java.util.ArrayList<>(List.of("133", "133.0a1"));
    List<String> b = new java.util.ArrayList<>(List.of("133.0a1", "133"));

    a.sort(new org.openqa.selenium.grid.data.SemanticVersionComparator());
    b.sort(new org.openqa.selenium.grid.data.SemanticVersionComparator());

    assertThat(a).isEqualTo(b);
```

Update: with valid input this returns good results, sorry for the confusion.

File: java/src/org/openqa/selenium/grid/data/SemanticVersionComparator.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    return 0; // Versions are equal
+  }
+
+  private boolean isNumber(String str) {
+    if (str == null || str.isEmpty()) {
+      return false;
+    }
+    for (char c : str.toCharArray()) {
+      if (!Character.isDigit(c)) {
```
Comment (joerg1985) [inline java/src/org/openqa/selenium/grid/data/SemanticVersionComparator.java]: There was a recent issue due to turkish locale, with this in mind:
`Character.isDigit` should return false positive results for none latin digits, from the javadoc:
```
    '\u0030' through '\u0039', ISO-LATIN-1 digits ('0' through '9')
    '\u0660' through '\u0669', Arabic-Indic digits
    '\u06F0' through '\u06F9', Extended Arabic-Indic digits
    '\u0966' through '\u096F', Devanagari digits
    '\uFF10' through '\uFF19', Fullwidth digits 
 ```

Comment (VietND96) [reply java/src/org/openqa/selenium/grid/data/SemanticVersionComparator.java]: @joerg1985, I am not sure the fix is good enough. I update to compare explicitly if out of range ISO-LATIN-1, it should be false.

**Review by diemol — APPROVED** [review]
Please fix https://github.com/SeleniumHQ/selenium/actions/runs/12504197675/job/34885613546?pr=14914#step:17:1469 and then we can merge.

**Review by diemol — APPROVED** [review]
